component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -683,4 +713,54 @@ private boolean validateJTIInCache(String jti, SignedJWT signedJWT, JWTCacheEntr
         }
         return true;
     }
+
+    /**
+     * Validate whether the request signing algorithm is configured for the application.
+     *
+     * @param requestSigningAlgorithm     The request signed algorithm.
+     * @param clientId                    Client ID of the application.
+     * @return whether the request signing algorithm is configured for the application.
+     * @throws OAuthClientAuthnException OAuth Client Authentication Exception.
+     */
+    private boolean isValidSignatureAlgorithm(String requestSigningAlgorithm, String clientId)
+            throws OAuthClientAuthnException {
+
+        //   Obtain the signing algorithm configured for the application.
+        List<String> configuredSigningAlgorithms = getConfiguredSigningAlgorithm(clientId);
+        //  Validate whether the JWT signing algorithm is configured for the application.
+        if (configuredSigningAlgorithms == null || configuredSigningAlgorithms.contains(requestSigningAlgorithm)) {
+            return true;
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("JWT signed algorithm does not match with the configured algorithms: " +
+                        configuredSigningAlgorithms);
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Obtain the request signing algorithms configured for the application.
+     *
+     * @param clientId   Client ID of the application.
+     * @throws OAuthClientAuthnException OAuth Client Authentication Exception.
+     */
+    private List<String> getConfiguredSigningAlgorithm(String clientId) throws OAuthClientAuthnException {
+
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+            for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                if (Constants.TOKEN_ENDPOINT_AUTH_SIGNING_ALG.equals(serviceProviderProperty.getName())) {
+                    return Arrays.asList(serviceProviderProperty.getValue());
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Error occurred while retrieving the service provider.",
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+        // null is returned when no algorithm is specified for the app and this will be handled accordingly.
+        return null;
```

Comment:
Lets use java optional instead

Reply:
Created an Arraylist at the beginning itself and assigned values to it in order to minimize the condition checks required. Fixed in e9d0e7eeb2260413659c89164eed65584a7d4278